Game/Plugins/MetricsServiceProvider/Source/MetricsServiceProvider/Private/MetricsServiceModule.cpp

```diff
@@ -0,0 +1,617 @@
+{
+	FAnalyticsProviderMetrics::Destroy();
+}
+
+TSharedPtr<IAnalyticsProvider> FAnalyticsMetricsServiceModule::CreateAnalyticsProvider(const FAnalyticsProviderConfigurationDelegate& GetConfigValue) const
+{
+	// Use our project name for our analytics endpoints.  By default UE4 uses the compiled build type of the game.
+	FString ProjectName;
+	FString AnalyticsSection;
+	if (FParse::Value(FCommandLine::Get(), TEXT("projectName"), ProjectName))
+	{
+		AnalyticsSection = TEXT("Analytics_") + ProjectName;
+	}
+	else
+	{
+		AnalyticsSection = TEXT("Analytics");
+	}
+
+	const FString Key = FAnalytics::Get().GetConfigValueFromIni(GEngineIni, AnalyticsSection, TEXT("ApiKey"), true);
+	const FString ApiEndpoint = FAnalytics::Get().GetConfigValueFromIni(GEngineIni, AnalyticsSection, TEXT("ApiEndpoint"), true);
+	const int32 BatchSize = FCString::Atoi(*FAnalytics::Get().GetConfigValueFromIni(GEngineIni, AnalyticsSection, TEXT("BatchSize"), true));
+	const FString BinaryKey = FAnalytics::Get().GetConfigValueFromIni(GEngineIni, AnalyticsSection, TEXT("BinaryApiKey"), true);
+	const FString BinaryEndpoint = FAnalytics::Get().GetConfigValueFromIni(GEngineIni, AnalyticsSection, TEXT("BinaryEndpoint"), true);
+	const float MaxAge = FCString::Atof(*FAnalytics::Get().GetConfigValueFromIni(GEngineIni, AnalyticsSection, TEXT("MaxAge"), true));
+
+	return FAnalyticsProviderMetrics::Create(Key, ApiEndpoint, BinaryKey, BinaryEndpoint, BatchSize, MaxAge);
+}
+
+FAnalyticsProviderMetrics::FAnalyticsProviderMetrics(const FString& Key, const FString& ApiEndpoint, const FString& BinaryKey, const FString& BinaryEndpoint, const int32 BatchSize, const float FlushPeriod) :
+	ApiKey(Key),
+	BinaryApiKey(BinaryKey),
+	EndPointURL(ApiEndpoint),
+	BinaryEndpointURL(BinaryEndpoint),
+	BatchSizeThreshold(BatchSize),
+	EngineVersion(*FEngineVersion::Current().ToString()),
+	EventEnvironment(LexToString(FApp::GetBuildConfiguration())),
+	FlushPeriodThresholdSeconds(FlushPeriod)
+{
+	HttpRetryManager = MakeShared<FHttpRetrySystem::FManager>(
+		FHttpRetrySystem::FRetryLimitCountSetting(),
+		FHttpRetrySystem::FRetryTimeoutRelativeSecondsSetting());
+
+	Http = &FHttpModule::Get();
+
+	Prometheus = MakeShared<FPrometheusServer>();
+	Prometheus->Initialize();
+
+	DeltaSecondsSinceFlush = 0.0f;
+
+	TelemetryReported = Prometheus->GetMetric(TEXT("telemetry_events"), TArray<FPrometheusLabel>{});
+
+	TelemetryClassEvent(TEXT("session"), TEXT("Metrics Started"), ProfileId);
+}
+
+FAnalyticsProviderMetrics::~FAnalyticsProviderMetrics()
+{
+	if (bHasSessionStarted)
+	{
+		EndSession();
+	}
+	TelemetryClassEvent(TEXT("session"), TEXT("Metrics Destructed"), ProfileId);
+}
+
+bool FAnalyticsProviderMetrics::Tick(float DeltaSeconds)
+{
+	HttpRetryManager->Update();
+
+	DeltaSecondsSinceFlush += DeltaSeconds;
+	if (DeltaSecondsSinceFlush >= FlushPeriodThresholdSeconds)
+	{
+		FlushEvents();
+		DeltaSecondsSinceFlush = 0.0f;
+	}
+
+	return true;
+}
+
+#if ENGINE_MINOR_VERSION >= 26
```

Comment:
I think the sexy way to do this would be to do something like:
#if ENGINE_MINOR_VERSION >= 26
using HttpRequest = TSharedRef<IHttpRequest, ESPMode::ThreadSafe>
#else
using HttpRequest = TSharedRef<IHttpRequest>

and then having just one function that uses HttpRequest.